modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -1978,13 +1980,109 @@ def sort_by(
+            if len(self.axes[0]) == 0 or len(self.axes[1]) == 0:
+                return self.copy()
+            # If this df only has one row partition, we don't want to do a shuffle and sort - we can
+            # just do a full-axis sort.
+            if len(self._partitions) == 1:
+                return self.apply_full_axis(
+                    1,
+                    sort_function,
+                )
+            if self.dtypes[columns[0]] == object:
+                # This means we are not sorting numbers, so we need our quantiles to not try
+                # arithmetic on the values.
+                method = "inverted_cdf"
+            else:
+                method = "linear"
+            from .utils import build_sort_functions
+
+            shuffling_functions = build_sort_functions(
+                self,
+                columns[0],
+                method,
+                ascending if not isinstance(ascending, list) else ascending[0],
+                **kwargs,
+            )
+            major_col_partition_index = self.columns.get_loc(columns[0])
+            cols_seen = 0
+            index = -1
+            for i, length in enumerate(self._column_widths_cache):
+                cols_seen += length
+                if major_col_partition_index <= cols_seen:
+                    index = i
+                    break
+            new_partitions = self._partition_mgr_cls.shuffle_partitions(
+                self._partitions,
+                index,
+                shuffling_functions.sample_function,
+                shuffling_functions.pivot_function,
+                shuffling_functions.split_function,
+                sort_function,
+            )
+            new_axes = self.axes
+            if kwargs.get("ignore_index", False):
+                new_axes[axis.value] = RangeIndex(len(new_axes[axis.value]))
+            else:
+                new_axes[axis.value] = self._compute_axis_labels_and_lengths(
+                    axis.value, new_partitions
+                )[0]
+            # Sometimes when performing a sort for `value_counts` the partitions have a MultiIndex
+            # but the _compute_axis_labels_and_lengths function gives us an Index of tuples instead
+            # of a MultiIndex.
+            if isinstance(self.axes[axis.value], pandas.MultiIndex) and isinstance(
+                new_axes[axis.value][0], tuple
+            ):
+                new_axes[axis.value] = pandas.MultiIndex.from_tuples(
+                    new_axes[axis.value].values
+                )
+            new_axes[axis.value] = new_axes[axis.value].set_names(
+                self.axes[axis.value].names
+            )
+            new_axes[axis.value ^ 1] = self.axes[axis.value ^ 1]
+            new_lengths = [None, None]
+            (
+                new_partitions,
+                new_lengths[axis.value],
+            ) = self._partition_mgr_cls.rebalance_partitions(new_partitions)
```

Comment:
why do we have to rebalance partitions we've just constructed for sorting?..

Reply:
This is kind of a fail-safe - if the pivots we pick are really poor, we can end up with skewed partitions, so it makes sense to throw a rebalance on the end just to ensure we don't skew partition sizes (although in practice, with the Cauchy distribution, our partitions always end up fairly balanced. I'm keeping this in case users end up having pathological distributions like only a couple of unique values that will introduce skew in the partitioning.)

Reply:
can we add some heuristic that would check if partitions are "mostly" balanced, and only rebalance if they're really skewed?

Reply:
We have a check like that [here](https://github.com/modin-project/modin/blob/9534478597c74a70943cba6fdde5f78f48a8fce9/modin/core/dataframe/pandas/partitioning/partition_manager.py#L1393-L1398) within `rebalance_partitions`. It's not checking for skew in partition sizes, but rather for too many partitions. We should always have roughly as many partitions as we started with, so  I don't think `rebalance_partitions` will do the kind of balancing you want. I think for this PR, it might be best to just add a note saying that we might want to fix skew, possibly by augmenting `rebalance_partitions`.

Reply:
I'm fine with adding a note or something, even with fixing this in a later PR.

It just feels a little too wasteful to rebalance partitions if pivot points were picked correctly, and all is already balanced (which should happen on random data pretty often).

Reply:
Sure, that makes sense to me!